[nativert] Add utility function to convert strings into numbers.

zhxchen17 opened the pull request:
Summary:

nativert RFC: https://github.com/zhxchen17/rfcs/blob/master/RFC-0043-torch-native-runtime.md

To land the runtime into PyTorch core, we will gradually land logical parts of the code into the Github issue and get each piece properly reviewed.

This diff adds a small library to convert strings into numbers which will later be used for parsing graph IR.

Differential Revision: D73133034

## Test Plan

c10 unittests

henryoier reviewed (commented):
Code change looks good to me. 

Seems the test plan is still using internal buck build targets, should we also add a OSS test command here?

malfet commented on torch/nativert/common/Conv.h:
```diff
@@ -0,0 +1,30 @@
+
+/*
+ * Convert a string to an integer. prefixes like "0x" or trailing whitespaces
+ * are not supported. Similayly, integer string with trailing characters like
```
```suggestion
 * are not supported. Similarly, integer string with trailing characters like
```

malfet commented on torch/nativert/common/Conv.h:
```diff
@@ -0,0 +1,30 @@
+
+#include <c10/macros/Export.h>
+
+namespace torch::nativert {
```
Why does it need to be in nativert namespace? `at::utils::` feels like a more appropriate namespace. And there probably already a function there

malfet commented on test/cpp/nativert/test_conv.cpp:
```diff
@@ -0,0 +1,31 @@
+namespace torch::nativert {
+
+TEST(TryToTest, Int64T) {
+  // Test valid conversions
```
At least one test for negatives? Also is `000` a valid string?

malfet commented on test/cpp/nativert/test_conv.cpp:
```diff
@@ -0,0 +1,31 @@
+
+TEST(TryToTest, Double) {
+  // Test valid conversions
+  EXPECT_EQ(tryTo<double>("123.45"), 123.45);
```
Do you want to test that integers are conveyable as doubles?

malfet commented on test/cpp/nativert/test_conv.cpp:
```diff
@@ -0,0 +1,31 @@
+TEST(TryToTest, Double) {
+  // Test valid conversions
+  EXPECT_EQ(tryTo<double>("123.45"), 123.45);
+  EXPECT_EQ(tryTo<double>("-123.45"), -123.45);
```
Would be nice to add tests for fractional numbers without leading zero(`.5`, `-.02`)  as well as scientific numerals `5e-2` or `1e+3`

albanD commented on torch/nativert/common/Conv.cpp:
```diff
@@ -0,0 +1,51 @@
+namespace torch::nativert {
+
+template <>
+std::optional<int64_t> tryTo<int64_t>(std::string_view symbol) {
```
I would expect fmt has such functions?
@swolchok or @cyyever might know off the top of their head

swolchok replied:
[std::stoll](https://en.cppreference.com/w/cpp/string/basic_string/stol) seems likely to be nicer than making use of bare strtoll. There is also [std::from_chars](https://en.cppreference.com/w/cpp/utility/from_chars) but apparently libc++ didn't support it as recently as last year. (Disclaimer: I haven't used either of these, but I would try.)

zhxchen17 replied:
@swolchok I tried std::from_chars but ran into some weird compile errors from MSVC. :( I can give it another try on stoll.

zhxchen17 replied:
Hmm I'm still inclined to use strtoll. std::stoll seems to throw exception and we expect the nullopt case for this function.

cyyever commented:
I would use `std::to_string` or libfmt or something else under `c10/utils/string*`. If these functions are indeed irreplaceable, they should be move to `c10/utils/`

swolchok commented on test/cpp/nativert/test_conv.cpp:
```diff
@@ -0,0 +1,31 @@
+  EXPECT_EQ(tryTo<int64_t>("123"), 123);
+  EXPECT_EQ(tryTo<int64_t>("456"), 456);
+  // Test invalid conversions
+  EXPECT_FALSE(tryTo<int64_t>("0x123").has_value());
+  EXPECT_FALSE(tryTo<int64_t>("123abc").has_value());
+  EXPECT_FALSE(tryTo<int64_t>("123.45").has_value());
+  EXPECT_FALSE(tryTo<int64_t>("").has_value());
+  // Test overflow
+  EXPECT_FALSE(tryTo<int64_t>("12345678901234567890").has_value());
```
you could make this less verbose by putting the test strings in a vector and looping:

```
const std::vector<const char*> invalid_examples = {
  "0x123",
  "123abc",
  ...
};
for (const char* bad : invalid_examples) {
  EXPECT_FALSE(tryTo<int64_t>(bad).has_value());
}
```

(and similarly for the valid examples and other test cases)

swolchok commented on torch/nativert/common/Conv.cpp:
```diff
@@ -0,0 +1,51 @@
+std::optional<int64_t> tryTo<int64_t>(std::string_view symbol) {
+  // TODO Using strtoll for portability. Consider using std::from_chars in the
+  // future.
+  std::string symbol_str(symbol);
```
since we need to copy the input if we don't have a guaranteed null terminator, I would provide efficient overloads of this function taking `const std::string&` and `const char*`. It should be easy to extend your tests to cover the overloads as well if you use the array-based approach I outlined.

swolchok commented on torch/nativert/common/Conv.h:
```diff
@@ -0,0 +1,30 @@
+
+/*
+ * Convert a string to a double. prefixes like "0x" or trailing whitespaces
+ * are not supported. Similayly, integer string with trailing characters like
```
Similarly misspelled here also

swolchok commented on torch/nativert/common/Conv.h:
```diff
@@ -0,0 +1,30 @@
+TORCH_API std::optional<int64_t> tryTo<int64_t>(std::string_view symbol);
+
+/*
+ * Convert a string to a double. prefixes like "0x" or trailing whitespaces
```
why do we assume all consumers need a full 64 bits of precision?

zhxchen17 replied:
Originally this function was used to parse textual format FX graph from c++ (in nativert), where we always expected double precision.

swolchok commented on torch/nativert/common/Conv.cpp:
```diff
@@ -0,0 +1,51 @@
+template <>
+std::optional<int64_t> tryTo<int64_t>(std::string_view symbol) {
+  // TODO Using strtoll for portability. Consider using std::from_chars in the
+  // future.
```
nit: would be nice to link out to https://libcxx.llvm.org/Status/Cxx17.html stating that from_chars isn't fully supported until clang 20 (though for integers you could potentially use it now if MSVC versions have support)

swolchok commented on torch/nativert/common/Conv.cpp:
```diff
@@ -0,0 +1,51 @@
+}
+
+template <>
+std::optional<double> tryTo<double>(std::string_view symbol) {
```
same comments as above implementation

zhxchen17 commented:
Moved to function to c10/util/

Added more tests and overloads for std::string.

albanD commented on c10/util/StringUtil.h:
```diff
@@ -213,6 +213,31 @@ inline void printQuotedString(std::ostream& stmt, const std::string_view str) {
+/*
+ * Convert a string to a double. Trailing whitespaces are not supported.
+ * Similarly, integer string with trailing characters like "123abc" will
+ * be rejected either.
```
```suggestion
 * be rejected.
```

albanD commented on c10/util/StringUtil.h:
```diff
@@ -213,6 +213,31 @@ inline void printQuotedString(std::ostream& stmt, const std::string_view str) {
+/*
+ * Convert a string to a 64 bit integer. Trailing whitespaces are not supported.
+ * Similarly, integer string with trailing characters like "123abc" will be
+ * rejected either.
```
```suggestion
 * rejected.
```

albanD commented on c10/util/StringUtil.h:
```diff
@@ -213,6 +213,31 @@ inline void printQuotedString(std::ostream& stmt, const std::string_view str) {
 }
 
+template <typename T>
+std::optional<T> tryTo(const char* symbol) = delete;
```
```suggestion
std::optional<T> tryToNumber(const char* symbol) = delete;
```

albanD commented on c10/util/StringUtil.cpp:
```diff
@@ -146,4 +146,52 @@ size_t ReplaceAll(std::string& s, std::string_view from, std::string_view to) {
+
+template <>
+std::optional<int64_t> tryTo<int64_t>(const char* symbol) {
+  // TODO Using strtoll for portability. Consider using std::from_chars in the
```
I guess this is outdated now?

zhxchen17 replied:
we are still using strtoll because there's no better alternative atm. When we upgrade to newer clang and msvc we should give it another try as the comments suggested here.

albanD commented on c10/util/StringUtil.cpp:
```diff
@@ -146,4 +146,52 @@ size_t ReplaceAll(std::string& s, std::string_view from, std::string_view to) {
+
+template <>
+std::optional<double> tryTo<double>(const char* symbol) {
+  // TODO Using strtod for portability. Consider using std::from_chars in the
```
Also outdated?

albanD commented on c10/test/util/string_util_test.cpp:
```diff
@@ -77,4 +77,58 @@ TEST(StringUtilTest, testStrMulti) {
 }
 } // namespace test_str_multi
 
+namespace test_try_to {
```
I'm not sure how the C++ tests work. Do you know which shard and when this is executed in CI ?

zhxchen17 replied:
@albanD I don't know how I can find the exact shard, but when I had a bug in unittest, the CI is able to turn red for some shard... (I couldn't find it anymore since it's overwritten). 

Weirdly we only have 4 checks run on this PR now with the new push.

albanD replied:
Given the msvc/clang comment below, hopefully this runs on all 3 OS.
Also for future tests that might need cuda or rely on specific hardware. We will need to know exactly where it runs to make sure we have the appropriate coverage.

albanD reviewed (commented):
Ho sad that we can't do that using fmt directly :/

cyyever commented:
> Ho sad that we can't do that using fmt directly :/

It is possible to use libfmt if the implementation was moved to `.cpp`.

zhxchen17 commented:
> > Ho sad that we can't do that using fmt directly :/
> 
> It is possible to use libfmt if the implementation was moved to `.cpp`.

@cyyever We're adding a function converting from string to number, not the opposite direction.

cyyever commented:
For the ``const char *`` overloads, what happens when nullptr is used?

zhxchen17 committed:
[nativert] Add utility function to convert strings into numbers. (#151467)

Summary:

Torch Native Runtime RFC: <TODO>

To land the runtime into PyTorch core, we will gradually land logical parts of the code into the Github issue and get each piece properly reviewed.

This diff adds a small library to convert strings into numbers which will later be used for parsing graph IR.

Test Plan: c10 tests

Reviewed By: dolpm

Differential Revision: D73133034

cyyever commented:
I just checked cppreference that if nullptr is passed, the return value is 0. It's UB from semantics viewpoint but the behaviour is specified.

zhxchen17 commented:
> I just checked cppreference that if nullptr is passed, the return value is 0. It's UB from semantics viewpoint but the behaviour is specified.

Yeah I was referring a possible dereference of end pointer in that case. 

For some reason godbolt gave me SIGSEGV with nullptr: https://godbolt.org/z/Tx1YMhGr9

cyyever commented:
@zhxchen17 Then you have to check nullptr before calling strtoll

cyyever commented:
Should we overload `std::string_view`?

albanD reviewed (approved):
Ok!

zhxchen17 commented:
@pytorchbot merge
